kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,84 @@
+package kyo
+
+import Layers.internal.*
+import kyo.Flat.unsafe
+import kyo.core.*
+import scala.annotation.targetName
+
+sealed trait Layer[-In, +Out, -S]:
```

Comment:
I was hoping we could have only two type parameters but it doesn't seem possible with memoization. I think it's ok.

Reply:
It should be possible to encode it as:

```
Layer[S & Envs[In], Out]
```
If we can unpack the `Envs[A]`...

Reply:
Interesting! That'd be great